README.md
# Automated loess regression

In regression problems, a parametric model for the response (or dependent) variable as a function of the predictor (or independent) ones is not always available or known. In those cases, a non-parametric regression is highly recommended because it avoids the assumption of a heuristic model. However, the price to pay is that more data are needed to perform a fit, compared to the case of parametric regressions.

One of those non-parametric regressions is **loess** (Cleveland et al. 1992, chapter 8 of [Statistical Models in S](https://ui.adsabs.harvard.edu/abs/1992sms..book.....C/abstract)), which performs polynomial fits over local intervals along the domain. To perform a loess regression with one predictor, it is necessary to specify the order of the local polynomial (deg), which can be 1 (linear) or 2 (quadratic), and the smoothing parameter (alpha), which determines how much of the data is used to fit each local polynomial.

In [Rodríguez et al. (2019)](https://ui.adsabs.harvard.edu/abs/2019MNRAS.483.5459R/abstract), I developed a routine to perform **loess** regressions to data with one predictor variable, called **Automated Loess Regression (ALR)**. This routine takes into account the observed and intrinsic errors, along with the presence of possible outliers. The routine uses by default an order 2 for the local polynomials in order to give more freedom to the **loess** fitting procedure. In order to estimate an optimal alpha value, **ALR** uses the "an" information criterion (AIC, [Akaike 1974](https://doi.org/10.1109/TAC.1974.1100705)). To test whether response errors can account for the observed dispersion around the **ALR** fit, the code computes its log-likelihood. If an intrinsic error is necessary to maximize the log-likelihood, then it is added in quadrature to the response errors and the code performs again the **ALR** fit. This process is repeated until an intrinsic error is not necessary. For the outliers detection, **ALR** uses the [Tukey (1977)](https://www.pearson.com/us/higher-education/product/Tukey-Exploratory-Data-Analysis/9780201076165.html) rule, where values below Q1-1.5(Q3-Q1) or above Q3+1.5(Q3-Q1) (known as inner fences, where Q1 and Q3 are the first and third quartile, respectively) are considered as outliers. Finally, the code computes the errors around the **ALR** fit through simulations.

When **loess** is not able to perform a regression (e.g., only few data points are available), then the **ALR** just performs a linear interpolation between points. 

I developed the **ALR** with the purpose of fitting [light curves](https://en.wikipedia.org/wiki/Light_curve) of [Type II supernovae](https://en.wikipedia.org/wiki/Type_II_supernova) during the first ~100 days. However, in principle, it can be applied to any set of data with one predictor variable, with or without errors on the response variable.

For any question, email me at olrodrig@gmail.com

**If you use the ALR in your work, please cite [Rodríguez et al. (2019)](https://ui.adsabs.harvard.edu/abs/2019MNRAS.483.5459R/abstract).**


CHANGELOG
2020/04/24: version 1.0
2020/09/01: version 1.1
2023/11/23: version 1.2


tests/test_ALR.R
source('../src/ALR.R')

#input data
x     <- c(0.5, 0.6, 0.7, 0.8, 0.9, 1.0, 1.1, 1.2, 1.3, 1.4)
y     <- c(0.1, 0.2, 0.6, 0.8, 0.9, 0.9, 1.3, 1.4, 1.7, 1.9)
err_y <- 0.1

ALR <- Automated_Loess_Regression(x, y, err_y=err_y)

#evaluate the ALR fit at xi=0.6
xi <- 0.6
print(ALR$interp(xi))

#print some ALR attributes
print(ALR$deg)   #degree of the local polynomials
print(ALR$alpha) #smoothing parameter
print(ALR$enp)   #equivalent number of parameters


tests/test_ALR.py
import sys
sys.path.append('../src')
from ALR import Automated_Loess_Regression

#input data
x     = [0.5, 0.6, 0.7, 0.8, 0.9, 1.0, 1.1, 1.2, 1.3, 1.4]
y     = [0.1, 0.2, 0.6, 0.8, 0.9, 0.9, 1.3, 1.4, 1.7, 1.9]
err_y = 0.1

ALR = Automated_Loess_Regression(x, y, err_y=err_y)

#evaluate the ALR fit at xi=0.6
xi = 0.6
print(ALR.interp(xi))

#print some ALR attributes
print(ALR.deg)   #degree of the local polynomials
print(ALR.alpha) #smoothing parameter
print(ALR.enp)   #equivalent number of parameters


src/ALR.py
import os
import numpy as np
import rpy2
from rpy2.robjects.packages import STAP
from rpy2.robjects.vectors import FloatVector
from plot_ALR import plot_ALR

class Automated_Loess_Regression:
  def __init__(self, x, y, err_y=0, deg=2, alpha=0, outliers_det=False, n_sims=1000, average=True, verbose=False):
    """
    Automated loess regression (ALR).
    
    Execute the R code ALR.R, which is an automated routine to perform loess regressions to data with one predictor variable. 
    This routine takes into account the observed and intrinsic errors, along with the presence of possible outliers.
    """
    
    deg, outliers_det = int(deg), bool(outliers_det)
    
    if hasattr(alpha, '__len__') == False:  alpha = float(alpha) 
    
    #Convert Python vectors to R vectors
    if hasattr(alpha, '__len__') == True:  alpha = FloatVector(alpha) 
    rx, ry, rerr_y = FloatVector(x), FloatVector(y), 0
    if hasattr(err_y, '__len__') == True:
        rerr_y = FloatVector(err_y)
    else:
        if err_y != 0:
            err_y  = np.linspace(err_y, err_y, len(y))
            rerr_y = FloatVector(err_y)
    
    #Run Automated_Loess_Regression in ALR.R
    path = os.path.dirname(__file__)
    with open(path+'/ALR.R', 'r') as f:
        string = f.read()
    try:    
        ALR=STAP(string,"Automated_Loess_Regression").Automated_Loess_Regression(rx, ry, err_y=rerr_y, deg=deg, alpha=alpha, outliers_det=outliers_det, average=average, n_sims=n_sims, verbose=verbose)
    
        ###ALR ATTRIBUTES###
            
        ###INPUT DATA
        self.deg           = deg
        self.outliers_det  = outliers_det
        self.n_data        = len(x)
        self.x             = x
        self.y             = y
        self.err_y         = 0
        self.with_y_errors = False
        if hasattr(err_y, '__len__') == True:  
            self.err_y = err_y
            self.with_y_errors = True
        
        ###OUTPUT DATA
        self.x_ALR      = np.array(ALR.rx2('x_ALR'))
        self.y_ALR      = np.array(ALR.rx2('y_ALR'))
        self.err_y_ALR  = np.array(ALR.rx2('err_y_ALR'))
        self.y_ALR_sims = np.array(ALR.rx2('y_ALR_sims'))
        self.alpha      = ALR.rx2('alpha')[0]
        self.enp        = ALR.rx2('enp')[0]
        self.ssd        = ALR.rx2('ssd')[0]
        self.err_0      = ALR.rx2('err_0')[0]
        self.n_outliers = int(ALR.rx2('n_outliers')[0])
        self.n_fit      = int(self.n_data - self.n_outliers)
        if self.n_outliers != 0:
            self.x_outliers     = np.array(ALR.rx2('x_outliers'))
            self.y_outliers     = np.array(ALR.rx2('y_outliers'))
            self.err_y_outliers = np.array(ALR.rx2('err_y_outliers'))
    except rpy2.rinterface_lib.embedded.RRuntimeError:
        print('Execution halted')
  
  #interpolate the ALR fit to the x values
  def interp(self, x):
        
      out_of_range = False
      if hasattr(x, '__len__') == True:
          if min(x) < min(self.x_ALR) or max(x) > max(self.x_ALR):  out_of_range = True
      else:
          if x < min(self.x_ALR) or x > max(self.x_ALR): out_of_range = True
      
      if out_of_range:
          print('ERROR (in interp): x has values out of the valid range (i.e., ['+str(min(self.x_ALR))+', '+str(max(self.x_ALR))+']).')
      else:    
          y_ALR     = np.interp(x, self.x_ALR, self.y_ALR)
          err_y_ALR = np.interp(x, self.x_ALR, self.err_y_ALR)
          return y_ALR, err_y_ALR
  
  #plot the ALR regression
  def plot(self, ALRs=[], comparison_func=[], invert_y_axis=False, object_name='input data', xlabel='x', ylabel='y', figure_name=''):
      if len(ALRs) == 0:  ALRs = [self]
      plot_ALR(ALRs, comparison_func=comparison_func, invert_y_axis=invert_y_axis, object_name=object_name, xlabel=xlabel, ylabel=ylabel, figure_name=figure_name)


tutorial/tutorial_routines.py
import numpy as np
    
def select_data_in_range(x, y, err_y, x_range):
  x_sec     = x[x >= min(x_range)]
  y_sec     = y[x >= min(x_range)]
  if hasattr(err_y, '__len__') == True:
      err_y_sec = err_y[x >= min(x_range)]
  y_sec     = y_sec[x_sec <= max(x_range)]
  if hasattr(err_y, '__len__') == True:  
      err_y_sec = err_y_sec[x_sec <= max(x_range)]
  x_sec     = x_sec[x_sec <= max(x_range)]
  
  if hasattr(err_y, '__len__') == True:
      return x_sec, y_sec, err_y_sec
  else:
      return x_sec, y_sec

#read example data
def read_example_data(input_file_name, x_range=[], with_y_errors=True):
    
  #read input data
  x     = np.genfromtxt(input_file_name, usecols=0)
  y     = np.genfromtxt(input_file_name, usecols=1)
  if with_y_errors == True:
      err_y = np.genfromtxt(input_file_name, usecols=2)
  
  x_ref = 0.0
  if len(x_range) != 0:
      
      if len(x_range) == 2:
          if x_range[1] == 'lowest_x':
              x_ref = min(x)
  
      if with_y_errors == True:
          x, y, err_y = select_data_in_range(x, y, err_y, np.array(x_range[0])+x_ref)
      else:
          x, y = select_data_in_range(x, y, 0, np.array(x_range[0])+x_ref)
  
  if with_y_errors == True:
      return x, y, err_y
  else:
      return x, y


src/plot_ALR.py
import numpy as np
import matplotlib as mpl
import matplotlib.pyplot as plt

def plot_outliers(ax, axr, ALR):
    
  if ALR.with_y_errors:  
      x_out, y_out, err_y_out = ALR.x_outliers, ALR.y_outliers, ALR.err_y_outliers
  else:
      x_out, y_out = ALR.x_outliers, ALR.y_outliers

  ax.plot(x_out, y_out, 'or',mec='k', zorder=2, label='outliers ('+str(len(x_out))+')')
  residuals_out = y_out - ALR.interp(x_out)[0]
  axr.plot(x_out, residuals_out, 'or',mec='k')


def plot_sigma_limits(ax, x, y, err_y, lw=1, zorder=1, labels=False):
    
  if labels == True:
      ax.plot(x, y-1.0*err_y, '--', color='lime'  , lw=lw, zorder=zorder, label='1$\sigma$ limit')
      ax.plot(x, y-3.0*err_y, '--', color='orange', lw=lw, zorder=zorder, label='3$\sigma$ limit')
  else:
      ax.plot(x, y-1.0*err_y, '--', color='lime'  , lw=lw, zorder=zorder)
      ax.plot(x, y-3.0*err_y, '--', color='orange', lw=lw, zorder=zorder)
  ax.plot(x, y+1.0*err_y, '--', color='lime'  , lw=lw, zorder=zorder)
  ax.plot(x, y+3.0*err_y, '--', color='orange', lw=lw, zorder=zorder)
  
def set_ticks(ax, ax_ref):
    
  x_majors = ax_ref.xaxis.get_majorticklocs()
  x_minor  = ((max(x_majors) - min(x_majors)) / float(len(x_majors)-1))/5.0  
  ax.xaxis.set_minor_locator(mpl.ticker.MultipleLocator(base=x_minor)) 
  
  y_majors = ax_ref.yaxis.get_majorticklocs()
  y_minor  =  ((max(y_majors) - min(y_majors)) / float(len(y_majors)-1))/4.0  
  ax.yaxis.set_minor_locator(mpl.ticker.MultipleLocator(base=y_minor))
  
  ax.tick_params('both', length=6, width=1, which='major', direction='in')
  ax.tick_params('both', length=3, width=1, which='minor', direction='in')
  ax.xaxis.set_ticks_position('both')
  ax.yaxis.set_ticks_position('both')
    
#plot the input data and the loess fit, along with the residuals
def plot_ALR(ALRs, comparison_func=[], invert_y_axis=False, object_name='input data', xlabel='x', ylabel='y', figure_name=''):

  mpl.rcParams['legend.handlelength'] = 1.5
  plt.rc('axes', labelsize=12.5)
  plt.rc('xtick', labelsize=12.5)
  plt.rc('ytick', labelsize=12.5)
  plt.rc('legend', fontsize=12.5)
  
  axs, axrs = [], []
  xminmax, yminmax, rminmax = [], [], []
  if len(ALRs) == 1:
      fig = plt.figure(figsize=(8, 8))
      fig.subplots_adjust(left=0.105, bottom=0.057, hspace=0.01, right=1.00-0.01, top=0.92)
      axs.append(fig.add_subplot(211))
      axrs.append(fig.add_subplot(212))
  if len(ALRs) == 2:
      fig = plt.figure(figsize=(12, 8))
      fig.subplots_adjust(left=0.105*8.0/12.0, bottom=0.057, hspace=0.01, right=1.00-0.01*8.0/12.0, top=0.92, wspace=0.01)
      axs.append(fig.add_subplot(221))
      axrs.append(fig.add_subplot(223))
      axs.append(fig.add_subplot(222))
      axrs.append(fig.add_subplot(224))
  if len(ALRs) == 3:
      fig = plt.figure(figsize=(14, 8))
      fig.subplots_adjust(left=0.105*8.0/14.0, bottom=0.057, hspace=0.01, right=1.00-0.01*8.0/14.0, top=0.92, wspace=0.01*12.0/14.0)
      axs.append(fig.add_subplot(231))
      axrs.append(fig.add_subplot(234))
      axs.append(fig.add_subplot(232))
      axrs.append(fig.add_subplot(235))
      axs.append(fig.add_subplot(233))
      axrs.append(fig.add_subplot(236))
      
  for i in range(0, len(ALRs)):
      
      ALR = ALRs[i]
      ax, axr = axs[i], axrs[i]
      
      x             = ALR.x
      y             = ALR.y
      err_y         = ALR.err_y
      with_y_errors = ALR.with_y_errors
      N_data        = ALR.n_data
      N_fit         = ALR.n_fit
      N_outliers    = ALR.n_outliers
      alpha         = ALR.alpha
      deg           = ALR.deg
      outliers_det  = ALR.outliers_det
      
      
      #ALR fit
      x_ALR, y_ALR, err_y_ALR = ALR.x_ALR, ALR.y_ALR, ALR.err_y_ALR
      
      #plot input data
      if with_y_errors:  ax.plot([x,x], [y-err_y, y+err_y], '-', color='gray', zorder=1)
      ax.plot(x, y, 'o', color='white',mec='k', zorder=2, label=object_name+' ('+str(N_data)+')')
      
      #plot comparison functions
      if len(comparison_func) != 0:
          
          plots, labels = [], []
          for func in comparison_func:
              
              mask = (func['x']>=min(x_ALR)) & (func['x']<=max(x_ALR))
              x_res         = func['x'][mask]
              y_for_res     = func['y'][mask]
              if 'err_y' in func.keys():  err_y_for_res = func['err_y'][mask]
              
              #mask = x_res<=max(x_ALR)
              #y_for_res     = y_for_res[mask]
              #if 'err_y' in func.keys():  err_y_for_res = err_y_for_res[mask]
              #x_res         = x_res[mask]
              
              res = y_for_res - np.interp(x_res, x_ALR, y_ALR)
              
              plot_func       , = ax.plot(func['x'], func['y'], '-' , color=func['color'], zorder=3, lw=2)
              axr.plot(x_res, res, '-', color=func['color'], zorder=3, lw=2)
              plots.append(plot_func)
              labels.append(func['label'])
              if 'err_y' in func.keys():
                  plot_func_3sigma, = ax.plot(func['x'], func['y']-3.0*func['err_y'], '--' , color=func['color'], zorder=2, lw=2)
                  ax.plot(func['x'], func['y']+3.0*func['err_y'], '--' , color=func['color'], zorder=2, lw=2)
                  axr.plot(func['x'], res-3.0*func['err_y'], '--', color=func['color'], zorder=3)
                  axr.plot(func['x'], res+3.0*func['err_y'], '--', color=func['color'], zorder=3)
                  plots.append(plot_func_3sigma)
                  labels.append('3$\sigma$ limit')
          leg2 = ax.legend(plots, labels, loc='best')
      
      #label for ALR
      if alpha != 0:
          if outliers_det == True:
              if N_outliers != 0:
                  label_text = r'ALR(N='+str(N_fit)+'; alpha='+str(round(alpha,3))+')'
                  if deg == 1: label_text = r'ALR(N='+str(N_fit)+'; alpha='+str(round(alpha,3))+'; deg='+str(deg)+')'
              else:
                  label_text = r'ALR(N='+str(N_fit)+'; alpha='+str(round(alpha,3))+'; outliers_det=True)'
                  if deg == 1: label_text = r'ALR(N='+str(N_fit)+'; alpha='+str(round(alpha,3))+'; deg='+str(deg)+'; outliers_det=True)'
          else:
              if len(ALRs) == 3:
                  label_text = r'ALR('+str(N_fit)+r'; $\alpha$='+str(round(alpha,3))+'; outliers_det=F)'
                  if deg == 1: label_text = r'ALR('+str(N_fit)+r'; $\alpha$='+str(round(alpha,3))+'; deg='+str(deg)+'; outliers_det=F)'
              else:
                  label_text = r'ALR(N='+str(N_fit)+'; alpha='+str(round(alpha,3))+'; outliers_det=False)'
                  if deg == 1: label_text = r'ALR(N='+str(N_fit)+'; alpha='+str(round(alpha,3))+'; deg='+str(deg)+'; outliers_det=False)'
      else:
          label_text = 'linear intpol between points'
          
      #plot ALR fit
      ax.plot(x_ALR, y_ALR, '-', color='b', zorder=4, lw=2, label=label_text)
      plot_sigma_limits(ax, x_ALR, y_ALR, err_y_ALR, lw=1, zorder=1, labels=True)
      
      #residuals
      axr.plot([min(x), max(x)], [0.0, 0.0], '--k')
      residuals = y - ALR.interp(x)[0]
      
      #plot residuals
      if with_y_errors:  axr.plot([x,x], [residuals-err_y, residuals+err_y], '-', color='gray', zorder=1)
      axr.plot(x, residuals, 'o', color='white',mec='k', zorder=2)
      plot_sigma_limits(axr, x_ALR, 0.0, err_y_ALR, lw=2, zorder=3, labels=False)
        
      #outliers
      if N_outliers != 0:  plot_outliers(ax, axr, ALR)

      if N_outliers == 0:
          ax.legend(bbox_to_anchor=(0.0, 1.085, 1.0, 0.10),loc='best', ncol=2, borderaxespad=0., handletextpad=0.2, mode='expand', numpoints=1)
      else:
          ax.legend(bbox_to_anchor=(0.0, 1.085, 1.0, 0.10),loc='best', ncol=3, borderaxespad=0., handletextpad=0.2, mode='expand', numpoints=1)
      if len(comparison_func) != 0: ax.add_artist(leg2)
      if i == 0:
          ax.set_ylabel(ylabel)
          axr.set_ylabel('residuals')
      axr.set_xlabel(xlabel)      
      
      #limits
      x_for_limits, y_for_limits, err_y_for_limits = x, y, err_y
      r_for_limits, err_y_ALR_for_limits = residuals, err_y_ALR
      
      xmin, xmax = min(x_for_limits), max(x_for_limits)
      if i == len(ALRs)-1:  xminmax.extend([xmin, xmax])
      
      if hasattr(err_y, '__len__') == True:
          ymin, ymax = min(y_for_limits-err_y_for_limits), max(y_for_limits+err_y_for_limits)
      else:
          ymin, ymax = min(y_for_limits), max(y_for_limits)
      if i == len(ALRs)-1:  yminmax.extend([ymin, ymax])    
      
      if hasattr(err_y, '__len__') == True:
          rmin, rmax = min([min(-3.0*err_y_ALR_for_limits),min(r_for_limits-err_y_for_limits)]), max([max(3.0*err_y_ALR_for_limits), max(r_for_limits+err_y_for_limits)])
      else:
          rmin, rmax = min([min(-3.0*err_y_ALR_for_limits),min(r_for_limits)]), max([max(3.0*err_y_ALR_for_limits), max(r_for_limits)])
      if i == len(ALRs)-1:  rminmax.extend([rmin, rmax])
      
      ax.xaxis.set_major_formatter(mpl.ticker.NullFormatter())
      if i != 0: 
          ax.yaxis.set_major_formatter(mpl.ticker.NullFormatter())
          axr.yaxis.set_major_formatter(mpl.ticker.NullFormatter())
      
      
  xmin, xmax = min(xminmax), max(xminmax)
  dx = (xmax-xmin)*0.025
  ymin, ymax = min(yminmax), max(yminmax)
  dy = (ymax-ymin)*0.05
  rmin, rmax = min(rminmax), max(rminmax)
  dr = (rmax-rmin)*0.05
  
  for i in range(0, len(ALRs)):
      axs[i].set_xlim(xmin-dx, xmax+dx)
      axrs[i].set_xlim(xmin-dx, xmax+dx)
      axs[i].set_ylim(ymin-dy, ymax+dy)
      axrs[i].set_ylim(rmin-dr, rmax+dr)
      set_ticks(axs[i],axs[len(ALRs)-1])
      set_ticks(axrs[i],axrs[len(ALRs)-1])
      if invert_y_axis == True:
          axs[i].invert_yaxis()
          axrs[i].invert_yaxis()
  
  if figure_name != '':  fig.savefig(figure_name+'.pdf', dpi=100)
